### PR TITLE
blog: pmemkv's bindings description

### PR DESCRIPTION
**Pull request by lukaszstolarczuk**
preview: https://lukaszstolarczuk.github.io/2020/03/04/pmemkv-bindings.html

before merge:
- [x] pmemkv-python released
- [x] python's and nodejs' docs are available on https://pmem.io

---
This change is [Reviewable](https://reviewable.io/reviews/pmem/pmem.github.io/126)

### CODE REVIEW

**Comment by karczex** [comment]
___
*[_posts/2020-02-15-pmemkv-bindings.md, line 80 at r2](https://reviewable.io/reviews/pmem/pmem.github.io/126#-M02dQ5M3qaz6yf-fJTs:-M0SKNDr-G7j810V3AZW:bozphpd) ([raw file](https://github.com/pmem/pmem.github.io/blob/5516e0b5653c75e2a372f693f8ad8465c331196c/_posts/2020-02-15-pmemkv-bindings.md#L80)):*
<details><summary><i>Previously, lukaszstolarczuk (Łukasz Stolarczuk) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

It's not returning numeric value for OK status. Most functions doesn't have return statement at all.

**Comment by szyrom** [comment]
Imho this should be merged after pmemkv-python release AND pmemkv-python blogpost (@karczex ) AND after the merge of nodejs bindings support for direct access to pmem

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+### Common for bindings
+
+There are few common characteristics regarding all bindings:
+
+#### Work with libpmemkv 1.0
+
+All of our currently released bindings work properly with libpmemkv 1.0. Unfortunately their API
+may not be fully functionally equal to the native C/C++ API. These **bindings, which are
+in version 1.0** deliver API in accordance with **pmemkv 1.0**. In general, pmemkv's binding
+released in version X.Y is (always) functionally greater or equal to pmemkv in the same X.Y version.
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: Drop "in general".

I don't like the "functionally greater or equal". "Is compatible with pmemkv x.y"?

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+Additionally, if a binding's version is prior to 1.0 (any 0.x version), it should be considered
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: Bindings with major version 0 should be considered experimental.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+an experimental implementation.
+
+Taking into consideration that libpmemkv is developing in the fast pace, there are always
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: "The pmemkv project is currently undergoing heavy development, and it's likely that the core library feature set will outpace features of the various bindings."

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+some gaps in the current versions of bindings' implementations. Even more, each binding
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: Moreover, individual bindings are developed independently. We recommend checking with documentation for specific bindings to see if a feature is present.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+is developed in its own pace, so the best way to check compatibility with pmemkv
+is to read changelog of the latest release of a specific binding.
+
+#### Handling pmemkv statuses
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: This paragraph adds nothing to the post. In effect, what you are saying is that "there are some error states, they are handled differently for every language, see docs to check how".

I'd change this to something like "Idiomatic language support" in which you explain that we tried to make bindings feel natural to the language (idiomatic), and because of this we had to slightly divert from the original library in some areas, such as return values, error handling and so on.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+Every language supports handling statuses returned from pmemkv's API. Each has its own way,
+either translating each status into numeric value or wrapping them into exceptions. Bindings
+using exceptions do not use status "OK", just returns nothing or a numeric value
+(of `PMEMKV_STATUS_OK`) if no error occurred. Return value differs also for various functions -
+to see details read binding's documentation or pmemkv's man pages.
+
+#### Not always direct access to data
+
+**Except in Python** (and soon [in Node.js][nodejs_pr]), our bindings do not (yet)
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: Invert this.

The bindings listed below, unless otherwise noted, do not support ...

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+provide direct access to the data stored in the Persistent Memory.
+They need to use buffer to copy the data to.
+
+#### Examples, tests and CI
+
+Each repository is equipped with basic, needed features like:
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: Each repository includes examples, tests and CI.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+* examples, showing basic library usage,
+* tests, verifying current implementation,
+* Continuous Integration system, to check all incoming changes.
+
+#### Packaging specific for language
+
+Each language is different, so we try our best to use packaging and installation systems
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: remove "try our best", just say "we use"

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+native to (or one of the most common for) each binding's language.
+
+#### Contributions are welcome
+
+As all of pmem's organization repositories, the ones of pmemkv-bindings are also
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: Well, it's github, so unless we used a non-free license, it's a given that projects in pmem organization are open source.
I'd rephrase this to say that "pmemkv is BSD, so we welcome contributions."

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+open source and in each of them any support is more than welcome.
+
+### pmemkv-nodejs
+
+Node.js binding is the first one to be released in stable version (1.0). It means
+it guarantees backward compatibility and it provides API functionally equal to
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: This guarantees backward compatibility and stable API, compatible with libpmemkv 1.0.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+the native pmemkv C/C++ API **in version 1.0**.
+
+It also delivers API description in form of JSDoc documentation (accessible e.g. as html in
+[here](nodejs_pmem_io)) and NodeJS-native support for configuring engines by JSON object.
+It handles returned statuses as exceptions and does not return any value when no errors occur.
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: It uses exception for error handling.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+Installation of this project is supported by simply typing `npm install` within the source directory.
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: ... uses npm for distribution, and can be installed using with `npm install pmemkv-python`.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+It then installs the dependencies listed in package.json in the local node_modules folder.
+
+For more information, like required dependencies, usage and full installation guide,
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: such as

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 14]
```diff
@@ -0,0 +1,155 @@
...
+see [pmemkv-nodejs github repository][nodejs_gh].
+
+### pmemkv-python
+
+This is the second binding to be released in version 1.0. It has all functionalities
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: Rephrase this to match whats written for nodejs.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 15]
```diff
@@ -0,0 +1,155 @@
...
+To install this project in your system it's required to type in: `python3 setup.py install [--user]`,
+and then it is installed in a system path or (when optional parameter `--user` is given) within
+user's files (`~/.local/lib/`).
+
+For more information, like required dependencies, usage and full installation guide,
+see [pmemkv-python github repository][py_gh].
+
+### pmemkv-java
+
+Latest released version is 0.9 and it's working just fine with libpmemkv 1.0, but
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: and is compatible with libpmemkv 1.0, but comes with no guarantees of stable API.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 16]
```diff
@@ -0,0 +1,155 @@
...
+its API may still change. In this binding's code, the JNI framework is used,
+with implementation of pmemkv's C API delivered in separate repository.
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: This binding uses JNI to interoperate with C, and the pmemkv C JNI implementation is delivered in ...

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 17]
```diff
@@ -0,0 +1,155 @@
...
+It handles returned statuses as exceptions and does not return any value when no errors occur.
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: same as above.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 18]
```diff
@@ -0,0 +1,155 @@
...
+Installation process, when pmemkv and pmemkv-jni are installed in the system, comes down to
+typing, within source directory, command: `mvn install`.
+
+For more information, like required dependencies, usage and full installation guide,
+see [pmemkv-java github repository][java_gh]. To see the repository of JNI content,
+see [pmemkv-jni github repository][jni_gh].
+
+### pmemkv-ruby
+
+It's currently released as 0.9 version, properly working with pmemkv 1.0,
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: rephrase this to use the same language as above.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 19]
```diff
@@ -0,0 +1,155 @@
...
+### Looking forward
+
+In future more bindings may come to life. There are currently two (of which we are aware)
+still in development and outside of our organization, you can read about them here:
+* [Rust binding][rust_binding_iss]
+* [Go binding][go_binding_iss]
+
+If you feel like we're missing some language, that may be useful to have in our portfolio,
+or you just want to let us know we don't have information about some (possibly yours)
+pmemkv's binding, or your feel like we're missing some functionalities, don't hesitate to
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: `please file an issue to let us know. We are always eager to help our community grow.` or something like that.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 20]
```diff
@@ -0,0 +1,155 @@
...
+| Language  | Version | URL |
+| --------- | ------- | --- |
+| **Node.js** | **1.0** | **[GitHub][nodejs_gh]**, **[docs][nodejs_pmem_io]** |
+| **Python** | **1.0** | **[GitHub][py_gh]**, **[docs][py_pmem_io]** |
+| Java | 0.9 | [GitHub][java_gh] |
+| Ruby | 0.9 | [GitHub][ruby_gh] |
+
+### Looking forward
+
+In future more bindings may come to life. There are currently two (of which we are aware)
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: What are you trying to say?
That we are thinking about creating new bindings or that someone else can create bindings? I'm guessing both, but this isn't very clear to me.

File: _posts/2020-02-15-pmemkv-bindings.md [hunk 21]
```diff
@@ -0,0 +1,155 @@
...
+still in development and outside of our organization, you can read about them here:
```
Comment (pbalcer) [inline _posts/2020-02-15-pmemkv-bindings.md]: What's "outside of our organization"?. Say either "outside of core pmemkv team" or "developed by our community". This is an open source project.

"There are ... community developed bindings".

**Comment by lukaszstolarczuk** [comment]
We've merged the node.js PR with direct data access, so I've updated the content already. I've also changed the directory of the post (along with the post date update), so the review is aligned to the old file, sorry for the inconveniences.
